src/TorchSharp/DataLoader.cs

```diff
@@ -112,16 +123,51 @@ public bool MoveNext()
                             DisposeCurrent();
                             if (!MoveNextValue()) return false;
                             List<Dictionary<string, Tensor>> dic = new();
-                            dic.Add(dataset.GetTensor(currentVal));
-                            for (var i = 1; i < batchSize; i++) {
-                                if (!MoveNextValue()) break;
-                                dic.Add(dataset.GetTensor(currentVal));
-                            }
+                            var taskedBatchCount = 0;
+                            var taskBatchLock = new object();
+
+                            TaskBatch();
+                            var t = currentVal;
+
+                            //Run Async
+                            foreach(var _ in Enumerable.Range(1, num_worker - 1))
+                                ThreadPool.QueueUserWorkItem(CreateBatch);
+
+                            dic.Add(dataset.GetTensor(t));
+                            CreateBatch(null);
+
+                            while (dic.Count < taskedBatchCount) { } //Wait for every task finished
```

Comment:
If I see this correctly, this loop's behavior is barely predictable.

To see the point, imagine by the time this thread reaches this line, none of the `QueueUserWorkItem` has started their execution.

I believe in this case the while loop's condition will be `false`, so it won't wait for anything asynchronous.